src/input/json_to_storage.cc

```diff
@@ -93,106 +231,362 @@ bool JSONPath::get_ref_from_head(string & ref_address)
+                << JSONToStorage::EI_ErrorAddress(this) << JSONToStorage::EI_Format("YAML") << JSONToStorage::EI_JSON_Type( get_node_type() )
+             );
+	}
+	return 0;
+}
+
+
+double PathYAML::get_double_value() const {
+	if (head()->IsScalar()) {
+		try {
+			return head()->as<double>();
+		} catch (YAML::Exception) {
+	        THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("The value should be 'YAML double', but we found: ")
+	                << JSONToStorage::EI_ErrorAddress(this) << JSONToStorage::EI_Format("YAML") << JSONToStorage::EI_JSON_Type( get_node_type() )
+	             );
+		}
+	} else {
+        THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("The value should be 'YAML double', but we found: ")
+                << JSONToStorage::EI_ErrorAddress(this) << JSONToStorage::EI_Format("YAML") << JSONToStorage::EI_JSON_Type( get_node_type() )
+             );
+	}
+	return 0.0;
+}
+
+
+std::string PathYAML::get_string_value() const {
+	if (head()->IsScalar()) {
+		try {
+			return head()->as<std::string>();
+		} catch (YAML::Exception) {
+	        THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("The value should be 'YAML string', but we found: ")
+	                << JSONToStorage::EI_ErrorAddress(this) << JSONToStorage::EI_Format("YAML") << JSONToStorage::EI_JSON_Type( get_node_type() )
+	             );
+		}
+	} else {
+        THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("The value should be 'YAML string', but we found: ")
+                << JSONToStorage::EI_ErrorAddress(this) << JSONToStorage::EI_Format("YAML") << JSONToStorage::EI_JSON_Type( get_node_type() )
+             );
+	}
+	return "";
+}
+
+
+std::string PathYAML::get_node_type() const {
+	switch (head()->Type()) {
+	  case YAML::NodeType::Null: return "YAML null";
+	  case YAML::NodeType::Scalar: return "other scalar type";
+	  case YAML::NodeType::Sequence: return "YAML sequence";
+	  case YAML::NodeType::Map: return "YAML map";
+	  default: return "undefined type";
+	}
+}
+
+
+bool PathYAML::get_record_key_set(std::set<std::string> &keys_list) const {
+	if ( head()->IsMap() ) {
+		for (YAML::const_iterator it=head()->begin(); it!=head()->end(); ++it) {
+			keys_list.insert( it->first.as<std::string>() );
+		}
+        return true;
+    }
+
+	return false;
+}
+
+
+int PathYAML::get_array_size() const {
+	if (head()->IsSequence()) {
+		return head()->size();
+	}
+	return -1;
+}
+
+
+bool PathYAML::is_map_type() const {
+	return head()->IsMap();
+}
+
+
+bool PathYAML::is_sequence_type() const {
+	return head()->IsSequence();
+}
+
+
+PathYAML * PathYAML::clone() const {
+	return new PathYAML(*this);
+}
+
+
+bool PathYAML::get_ref_from_head(string & ref_address)
+{
+	return false;
+}
+
+
+bool PathYAML::has_descendent_index(bool value_at_declaration) {
+	const Node & head_node = *( nodes_.back() );
+	if (head_node.Tag() == "!type") {
```

Comment:
This still requires key 'TYPE', but we want to use tag for type specification. So the interface in PathBase needs to be different.